### PR TITLE
feat(server): expose the cache eviction knobs, and warn where they cannot act

### PR DESCRIPTION
**Pull request by vahid-sohrabloo**
Five cache features merged recently, and **none of them could be switched on by anyone running `rostam-server`** — or by an embedder using the public `rostam` package:

- `RelocatingEviction`
- `RelocateReserveIntervalMs`
- `InPlaceSameSizeUpdate`
- `InPlaceSeqlockReads`
- `SieveVisitedBit`

Each existed only as a `cache.Config` field. None appeared in `rostam.CacheConfig`, and none had a server flag. They were reachable only by code constructing `cache.Config` directly — which is exactly what their benchmarks do, and why the gap went unnoticed.

This plumbs all five through `rostam.CacheConfig` and exposes them as server flags. **No defaults change**; every knob stays off.

## Most of them do nothing in most deployments

Exposing a flag the server accepts and then ignores is worse than not exposing it: an operator enables it, sees no effect, and concludes the feature is broken. So each knob was checked against the code for every deployment shape:

| knob | single node, in memory | single node, with a data dir | cluster |
|---|---|---|---|
| `RelocatingEviction` | works — at eviction and in the background reserve | works — at eviction only | no effect |
| `RelocateReserveIntervalMs` | works, with `RelocatingEviction` | no effect | no effect |
| `InPlaceSameSizeUpdate` | works | no effect | no effect |
| `InPlaceSeqlockReads` | works, with `InPlaceSameSizeUpdate` | no effect | no effect |
| `SieveVisitedBit` | works, with `RelocatingEviction` | works, with `RelocatingEviction` | no effect |

Why: replicated shards are forced to reject-writes, because evict-oldest diverges across replicas, and every knob here acts only under ring-buffer eviction. In-place updates are heap-only, since an mmap page is the durable copy. The background reserve is skipped on mmap shards.

**The cluster column is precise, not approximate.** Shards are forced to reject-writes only on the multi-node path, when `Peers` is non-empty. An embedder calling `NewEmbedded` with no peers gets ring-buffer eviction on mmap — the data-dir column. `rostam-server -cluster` always passes at least its own peer, so the server is always in the cluster column; running `-cluster -bootstrap` with only its own peer confirms it, logging the forced reject-writes line for every shard.

## Inert combinations warn at startup

`warnInertCacheKnobs` runs from both `NewDirect` and `NewEmbedded`, so embedders get the same warnings as server operators. It logs one `slog.Warn` per knob, with `option`, `flag`, `deployment` and `reason` fields.

- **A deployment reason takes precedence over a pairing reason.** Telling a cluster operator to also enable `-relocating-eviction` would not make anything work.
- **Pairing no-ops warn too:** `SieveVisitedBit` or the reserve interval without `RelocatingEviction`, and `InPlaceSeqlockReads` without `InPlaceSameSizeUpdate`.
- **Only what the caller actually set is judged,** so an unset flag never warns on a cluster.
- **Never a hard error.** Every one of these settings is harmless where it does nothing, and a config shared across a cluster and a single-node box is over-specified, not broken.

Confirmed against the real binary: a heap server with seqlock reads and the SIEVE bit printed two warnings; a `-data` server with in-place printed one; a cluster with all five printed five; a cluster with none printed nothing.

## `-in-place-seqlock-reads` leads with its caveat

Its read protocol is a deliberate data race on page bytes, validated after the fact by a version counter, and both of its concurrent tests skip under `-race` — so a green race lane does not cover it. That was documented inside `cache/`; as a server flag it is now operator-facing, so the flag's one-line help opens with that caveat, and the docs carry it in a `danger` box. A test pins the caveat to the `-h` summary so it cannot quietly move below the fold.

## Where it is plumbed

- **`rostam.CacheConfig`** gains the five fields. The reserve interval follows the existing `TTLSweepIntervalMs` convention: 0 keeps the default, negative disables, positive sets it.
- **`direct.go` / `embedded.go`:** cache-config construction is extracted into functions so a test can inspect exactly what the store opens; both share `applyEvictionKnobs`.
- **Flags** are listed under Storage in `-h`. The `-relocate-reserve-interval` default is read from `cache.DefaultConfig()`, so it cannot drift from the engine.
- **Environment variables** need no extra code, since every flag already maps to a `ROSTAM_*` variable; a test covers it.
- **The config file is deliberately unchanged.** `fileConfig` is documented as holding only settings that have no flag, so each setting lives in one place and there is no flag-versus-file precedence to define.
- **`mcp.go` and `llmproxy.go`** build their own internal stores; the knobs stay off there and no warnings fire.

## Tests

`TestCacheConfigEvictionKnobsReachDirectCache` and `TestCacheConfigEvictionKnobsReachEmbeddedCache` are the regression proof: they set all five on `CacheConfig` and assert each arrives in the `cache.Config` the store actually opens, and that a zero `CacheConfig` keeps the engine defaults. Written first, they failed to compile on `main` — `unknown field RelocatingEviction in struct literal of type CacheConfig`.

`TestInertCacheKnobWarnings` covers 28 cases across the matrix, including those that must stay silent. `TestInertCacheKnobReasonNamesTheRealCause` checks which reason is reported. `TestInPlaceWarningMatchesCacheBehaviour` checks the matrix against the engine rather than against itself: `Stats().InPlaceUpdates` rises exactly where no warning fires.

Flag tests cover parsing via command line and environment, with interval edge cases.

`go test -short -p 2 . ./cmd/rostam-server`, `GOARCH=386 CGO_ENABLED=0 go vet ./...` and `GOOS=windows go vet` on the changed packages all pass.

## Docs

A new "Cache eviction knobs" section in `docs/server/running.md` with the matrix, the reasons, a sample warning and the seqlock caveat; a pointer from `docs/kv/cache.md`; and a changelog entry.

One case is documented but not warned about: the background reserve also skips a shard with fewer than four pages, which a small memory budget spread over many shards can reach. Warning on it would mean duplicating an unexported constant from `cache/`.

---
## Summary by cubic
Five cache eviction knobs — `RelocatingEviction`, `RelocateReserveIntervalMs`, `InPlaceSameSizeUpdate`, `InPlaceSeqlockReads`, and `SieveVisitedBit` — existed only as `cache.Config` fields, so no server flag or `rostam.CacheConfig` field could set them. This plumbs all five through `rostam.CacheConfig` and exposes them as `rostam-server` flags; no defaults change, the four switches stay off, and the reserve interval stays inert without `-relocating-eviction`.

**Warnings for inert knobs**

- Most knobs do nothing on most deployments: cluster shards refuse writes at capacity instead of evicting, so none act there; the in-place pair acts on every same-size rewrite, full shard or not, but only on a heap shard; the background reserve is in-memory-only.
- `NewDirect` and `NewEmbedded` log one warning per knob that is set but cannot act on the store being built, including pairing no-ops like `SieveVisitedBit` without `RelocatingEviction`.
- Warnings, not errors: a config shared between a cluster and a single-node box is over-specified, not broken.

**Notes**

- The new flags are `-relocating-eviction`, `-relocate-reserve-interval`, `-in-place-same-size-update`, `-in-place-seqlock-reads`, and `-sieve-visited-bit`; each maps to a `ROSTAM_*` variable.
- `-in-place-seqlock-reads` is a deliberate data race on page bytes, validated by a version counter; race detection does not cover it, and the flag's help leads with that caveat.
- `-relocate-reserve-interval` follows `TTLSweepIntervalMs`: 0 keeps the library default, negative disables, positive sets the interval.
- The config file is deliberately unchanged — it holds only settings with no flag, and a test pins that it rejects these five keys.

<sup>Written for commit a034bad2972b167944f712d746397401ed3459c1. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rostamlabs/rostam/pull/145?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: docs/kv/cache.md [hunk 1]
```diff
@@ -19,6 +19,14 @@ to `cache.New`).
 | `Durable` | false | `msync` on commit boundaries (bounded by `MsyncIntervalMs`, default 100 ms); false = opportunistic OS flushing |
 | `Mlock` | false | pin the mmap into RAM (needs `ulimit -l` headroom; failure logs and continues) |
 
+The opt-in eviction knobs — `RelocatingEviction`, `RelocateReserveIntervalMs`,
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply docs/kv/cache.md:22]: Fixed in fab5477. Valid: the four switches are off, but `RelocateReserveIntervalMs` carries a non-zero library default that is inert without `RelocatingEviction`. `docs/kv/cache.md` now says the knobs "have no effect by default" and names that one. I made the same correction wherever the claim appeared: `running.md`, the changelog entry, the `CacheConfig` comment in `store.go`, and the comments on `applyEvictionKnobs` and `cacheKnobFlags`.

File: store.go [hunk 2]
```diff
@@ -1339,4 +1339,45 @@ type CacheConfig struct {
 	// multiplier is GC headroom rather than engine overhead; set GOMEMLIMIT to
 	// bound RSS independently of it.
 	MaxMemoryBytes int64
+
+	// The five fields below are opt-in eviction knobs, all off by default. Each maps
+	// onto the cache.Config field of the same name, whose doc has the full trade-off.
+	// MOST OF THEM DO NOTHING ON MOST DEPLOYMENTS: every one acts only on a shard that
+	// evicts at capacity, which a cluster shard never does (replication forces it to
+	// refuse writes instead), and the in-place pair acts only on an in-memory shard.
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply store.go]: Fixed in fab5477. Valid: `inPlaceEligible` requires the ring-buffer policy and a heap backing, not a full shard, so same-size rewrites happen in place from the first rewrite. The read lock (`needsReadLockForGet`) or the seqlock applies whenever the flag is on. The `store.go` summary now says none of the knobs acts on a refuse-writes shard (the policy replication forces), and that only `RelocatingEviction`, `RelocateReserveIntervalMs` and `SieveVisitedBit` depend on the shard actually evicting. It also says the in-place pair changes every same-size rewrite and how reads are taken, full or not, but only on an in-memory shard. The same over-broad sentence was in the cluster bullet of `running.md` and in `docs/kv/cache.md`, and both are corrected too.

File: docs/server/running.md [hunk 3]
```diff
@@ -135,6 +140,87 @@ grouped by the areas on this page; `-help-all` prints every description in full.
...
+- **`-data`: no background reserve.** On a file-backed shard the reserve's
+  copies and the page it frees could reach disk out of order, so a crash could
+  lose a record that would otherwise have survived. Relocation there runs only
+  at eviction time.
+- **The background reserve also needs at least four pages per shard.** A small
+  `max_memory` spread over many `-shards` can leave fewer, and such a shard
+  keeps no reserve; `-relocating-eviction` still works, at eviction time only.
+
+The server does not refuse a knob that has no effect — a config file shared by
+a cluster and a single-node box is over-specified, not broken — but it logs one
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply docs/server/running.md]: Fixed in fab5477. Valid, and the wording was worse than ambiguous: `fileConfig` holds only settings that have no flag and decodes with `DisallowUnknownFields`, so a config file naming one of these knobs fails to load and the server refuses to start. `running.md` now says it outright near the top of the section: these are flags and `ROSTAM_*` environment variables only, the `-config` file does not accept them, and a file that names one will not load. The later sentence now reads "the same flags or `ROSTAM_*` environment shared by a cluster and a single-node box". The changelog entry says the same, and I fixed the matching phrase in the `warnInertCacheKnobs` comment. `TestLoadFileConfigRejectsCacheEvictionKnobs` pins the behaviour: it checks that each knob, in snake_case and in field-name form, is rejected by `loadFileConfig`, so the docs cannot drift from it.
